tensorboard/webapp/widgets/line_chart_v2/sub_view/line_chart_axis_utils.ts

```diff
@@ -0,0 +1,99 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {Formatter} from '../lib/public_types';
+import {LinearScale} from '../lib/scale';
+
+export interface MinorTick {
+  value: number;
+  tickFormattedString: string;
+}
+
+// Major tick, unlike the minor tick, spans a range.
+export interface MajorTick {
+  // Start of the major tick range. An end is implicitly defined by next major tick while
+  // it can certainly change to explicitly define the end.
+  start: number;
+  tickFormattedString: string;
+}
+
+function getNumLeadingZerosInFractional(value: number): number {
+  const maybeExponential = value.toExponential().split('e-', 2);
+  if (maybeExponential.length === 2) {
+    return Number(maybeExponential[1]) - 1;
+  }
+  return 0;
+}
+
+export function getTicksForLinearScale(
+  scale: LinearScale,
+  formatter: Formatter,
+  maxMinorTickCount: number,
+  low: number,
+  high: number
+): {minor: MinorTick[]; major: MajorTick[]} {
+  const minorTickVals = scale.ticks([low, high], maxMinorTickCount);
+
+  const diff = Math.abs(high - low);
+  if (diff > 1e-3) {
+    return {
```

Comment:
From what I can reason out, this is the same return value for any other Scale that is not TemporalScale. Can you refactor the code so you only have one place that generates this?

Something like:

```
if (scale instance of TemporalScale) {
  // generate TemporalScale specific thing
} else if (scale instance of LinearScale and diff <= 1e-3) {
  // generate small LinearScale specific thing 
} else {
  // Generate default thing.
  const minorTickVals = scale.ticks(this.axisExtent, maxMinorTickCount);
  return {
      major: [],
      minor: minorTickVals.map((tickVal) => {
        return {
          value: tickVal,
          tickFormattedString: formatter.formatTick(tickVal),
        };
      }),
    };
}
```

It would also address what I think is some odd asymmetry in how you change behavior based on Scale type. You branch for LinearScale at ngOnChanges() but you branch for TemporalScale further down in call hierarchy at getMajorTickValues(). It's a bit difficult to follow.

Reply:
I initially had a code like that but the "small LinearScale specific thing" was not so small and it was just generally hard to follow the logic when done that way. In my trial, division by the scale type was the easiest to follow as long as all implementations conform to the interface.

Reply:
I think it's worth illustrating my suggestion:
https://github.com/tensorflow/tensorboard/compare/bb3ce0eaa...1b41b0c29

I find that significantly easier to read:
1. It's clear that different Scales are handled differently.
2. It's clearer that LinearScale and TemporalScale can (often?) generate the standard tick values but there are clear special cases where they do not.
3. There is less code duplication in generating the standard tick values. Actually, even stronger: It is nice that we don't use two entirely different ways to generate the same standard result.
4. There is less code duplication before and after generating the ticks: Only calculdate maxMinorTickCount once; only translate the results into this.minorTicks and this.majorTicks once.

Reply:
Yeah, that's not that far from my ideal situation; to clarify, I did not like the code when minor and major generations broken up and each requiring some switch statements. I did not find temporal and default scale generation to be complex to warrant the move but I definitely can see why that is correct. I will add tests to the utility promptly.